fem/hyperbolic.hpp

```diff
@@ -258,26 +424,215 @@ class RusanovFlux : public RiemannSolver
+    * @param[in] nor normal vector (not a unit vector) (dim)
+    * @param[in] Tr face element transformation
+    * @param[out] grad Jacobian of F(u⁻,u⁺,x) n
+    * side = 1:
+    *    ½J(u⁻,x)n + ½λ
+    * side = 2:
+    *    ½J(u⁺,x)n - ½λ
+    */
+   void Grad(int side, const Vector &state1, const Vector &state2,
+             const Vector &nor, FaceElementTransformations &Tr,
+             DenseMatrix &grad) const override;
+
+   /**
+    * @brief  Average normal numerical flux over the interval [u⁻, u⁺] in the
+    * second argument of the flux F̂(u⁻,u,x) n
+    * @note The average normal flux F̄ n is required to be implemented in
+    * FluxFunction::ComputeAvgFluxDotN()
+    * @note Systems of equations are treated component-wise
+    *
+    * @param[in] state1 state value (u⁻) of the beginning of the interval
+    * (num_equations)
+    * @param[in] state2 state value (u⁺) of the end of the interval
+    * (num_equations)
+    * @param[in] nor normal vector (not a unit vector) (dim)
+    * @param[in] Tr face element transformation
+    * @param[out] flux ½(F̄(u⁻,u⁺,x)n + F(u⁻,x)n) - ¼λ(u⁺ - u⁻)
+    * @return max(F(u⁺,x)n, F(u⁻,x)n)
+    */
+   real_t Average(const Vector &state1, const Vector &state2,
+                  const Vector &nor, FaceElementTransformations &Tr,
+                  Vector &flux) const override;
+
+   /**
+    * @brief  Jacobian of average normal numerical flux over the interval
+    * [u⁻, u⁺] in the second argument of the flux F̂(u⁻,u,x) n
+    * @note The average normal flux F̄ n is required to be implemented in
+    * FluxFunction::ComputeAvgFluxDotN() and the Jacobian of flux J n in
+    * FluxFunction::ComputeFluxJacobianDotN()
+    * @note Only the diagonal terms of the J n are considered, i.e., systems
+    * are treated as a set of independent equations
+    *
+    * @param[in] side gradient w.r.t the first (u⁻) or second argument (u⁺)
+    * @param[in] state1 state value (u⁻) of the beginning of the interval
+    * (num_equations)
+    * @param[in] state2 state value (u⁺) of the end of the interval
+    * (num_equations)
+    * @param[in] nor normal vector (not a unit vector) (dim)
+    * @param[in] Tr face element transformation
+    * @param[out] grad Jacobian of F̄(u⁻,u⁺,x) n
+    * side = 1:
+    *    ½(F̄(u⁻,u⁺,x)n - F(u⁻,x)n) / (u⁺ - u⁻) - ½J(u⁻,x)n + ¼λ
+    * side = 2:
+    *    ½(F(u⁺,x)n - F̄(u⁻,u⁺,x)n) / (u⁺ - u⁻) - ¼λ
+    */
+   void AverageGrad(int side, const Vector &state1, const Vector &state2,
+                    const Vector &nor, FaceElementTransformations &Tr,
+                    DenseMatrix &grad) const override;
+
+protected:
+#ifndef MFEM_THREAD_SAFE
+   mutable Vector fluxN1, fluxN2;
+   mutable DenseMatrix JDotN;
+#endif
+};
+
+/**
+ * @brief Godunov flux, a special case of Engquist-Osher flux,
+ *    F̂ n = min F(u)n    for u⁻ ≤ u⁺, u ∈ [u⁻,u⁺]
+ *    F̂ n = max F(u)n    for u⁻ > u⁺, u ∈ [u⁺,u⁻]
+ * @note The implementation assumes monotonous F(u,x) in u
+ */
```

Comment:
I'm not familiar with Godunov flux defined this way.  I've seen it widely used as being the choice for the numerical flux $\hat{F}(\vec{u_-},\vec{u_+},x)$ such that $\hat{F}(\vec{u_-},\vec{u_+},x) = F(\tilde{u}(\vec{u_-},\vec{u_+},x))$, where $\tilde{u}(\vec{u_-},\vec{u_+},x)$ is the Riemann solution evaluated at $x$ to the corresponding Riemann problem
$$\frac{\partial \vec{u}}{\partial t} + \frac{\partial}{\partial \tilde{x}} \vec{F}(\vec{u}) = \vec{0}$$
with initial condition $\vec{u} = \vec{u_-}$ at $\tilde{x} < x$ and $\vec{u} = \vec{u_+}$ for $\tilde{x} > x$.  This implementation looks to be specific for scalar equations with monotonic flux functions... perhaps this should be named `EnquistOsherFlux` to be consistent with terminology in, e.g., https://www.sciencedirect.com/science/article/pii/S009630031630621X (which is consistent with other works I'm familiar with)?

Reply:
It is the same definition (eq. 2.17 in the paper), just assuming monotonicity similarly to `RusanovFlux` discussed above, so min/max over the interval reduces to min/max of the endpoints. I am not sure what we will do for systems, so I would leave it as it is for now.

Reply:
It is not the same definition: Equation 2.17 is the Godunov flux *specifically* for a scalar system, which does not generalize to systems (also mentioned in the paper). This is similar to implementing a Newton solver that can only specifically for scalar equations and naming it `NewtonSolver`: a user will likely think they can use `NewtonSolver` for a nonlinear system of equations. If you are opposed to naming it `ScalarGodunovFlux` then you need to safeguard against users assuming they can apply it to systems (being explicit in the documentation that it's only for scalar equations, adding a verify or assert statement to check the dimension is 1, etc.).

Also, I'm not sure you need to assume monotonicity, i.e., this is the Godunov flux for a scalar conservation flux regardless if the flux is monotonic?  I could be wrong.

Reply:
The thing is that we do not have any Riemann solver for systems, as both assume monotonicity, yet there is example 18(p) using (originally one of) them for a system and working just fine and this framework has made it to a release of MFEM, so maneuvering is very limited. Moreover, it does not make sense to rename one of them when both apply the same assumption. Secondly, we may decide to extend these classes for systems of equations instead of using a different name, but it is hard to predict now. The best I can do is to put warning somewhere, I think 🤔 .
Without the assumption of monotonicity, you would have to scan somehow the whole interval and still you would have no guarantee, because flux is not necessarily polynomial, so just theoretically it could be a delta function (or its closest representable continuous projection) and you would need a (nearly) infinite number of evaluation points to find it. In other words, people always use some simplifying assumptions, which can be weaker for sure, but this is fine for my purposes and the basic examples of linear advection and Burgers, which are offered.

Reply:
Ok, I moved the warning from example 18 directly to the constructor of `GodunovFlux`.

Reply:
Example 18 was using the Rusanov flux as a numerical flux for a system in a very valid way.  This PR is introducing a class that implements the Godunov flux for a scalar system and then applies it component-wise to a system, which is invalid because that is not a Godunov flux, i.e., the flux function from the conservation law evaluated at the Riemann solution. I will not approve until this is addressed in the manner in which I have already requested.

Reply:
For what it's worth, the way Godunov flux is typically implemented is that it leverages an abstract Riemann solver, which is typically problem specific, to provide the Riemann solution at x/t = 0 to be used as the numerical flux, i.e., the $\vec{u}(x/t)$ in $\vec{f}(\vec{u}(x/t))$.  Here, you've hard-coded a Riemann solver for scalar conservation laws, effectively that $u(x/t)$ is either $u_-$ (right-moving Riemann solution), $u_+$ (left-moving Riemann solution), or such that $f'(u) = 0$ (rarefaction Riemann solution). The particular form you've chosen leverages the particular properties that determine whether the Riemann solution is right-moving, left-moving, or a rarefaction fan to get the form you're using.

For linear systems, exact Riemann solvers are typically supplied (see, e.g., [my work showing a modification to standard linear elasticity Riemann solver](https://link.springer.com/article/10.1007/s10915-017-0459-y)).  For nonlinear systems, approximate Riemann solvers have been carefully constructed for common nonlinear systems (with that work continuing).  Using the scalar conservation law Riemann solution for general nonlinear systems is widely considered **not valid** except perhaps some corner cases, otherwise there wouldn't be need for more work on approximate Riemann solvers. I believe this invalid application of a Riemann solver for scalar equations to the Euler equations (which has a rich history of carefully constructed & tailored Riemann solvers) is what @tarikdzanic was also worried about in [this comment](https://github.com/mfem/mfem/pull/4513#issuecomment-2386745754) and [this comment](https://github.com/mfem/mfem/pull/4513#issuecomment-2389660056).

Reply:
The flux F(u,x) must be monotonous in u, so it is guaranteed that max_u (F(u,x)) for u in [u-,u+] is the same as max(F(u-,x),F(u+,x)). In different notations it might be called differently, but here it makes sense this way.
You are right it does not work perfectly for Burgers, only works when the without a change of sign. That is limiting, but documented and with the current framework I do not see other options.
I am not saying this works fully generally for systems and that is why I moved the warning from example 18  (introduced based on the mentioned requests) directly to the constructor, warning it is a component-wise implementation.

Reply:
> The flux F(u,x) must be monotonous in u, so it is guaranteed that max_u (F(u,x)) for u in [u-,u+] is the same as max(F(u-,x),F(u+,x)).

Whether or not max_u (F(u,x)) for u in [u-,u+] is the same as max(F(u-,x),F(u+,x)) does not address the issue that you've hardcoded a Riemann solution into `GodunovSolver` that is only valid for scalar conservation laws.

> You are right it does not work perfectly for Burgers, only works without a change of the sign. 

No, I said the opposite: the Riemann solution that you've hardcoded into `GodunovSolver` is indeed the Riemann solution for Burgers equation, because Burgers equation has a flux that is convex.

> I am not saying this works fully generally for systems and that is why I moved the warning from example 18 (introduced based on the mentioned requests) directly to the constructor, warning it is a component-wise implementation.

This does not address my concern that a class titled `GodunovSolver` indicates that you've implemented a Godunov solver for system when the hardcoded Riemann solution is not valid for use on systems.

My recommendation remains that this PR not be merged until this issue is adequately addressed (suggestions for how to do so are found above).

Reply:
Ok, to unblock this, I generalized `RusanovFlux::Grad()` for systems, so we have at least one `NumericalFlux` for them. Hence, it makes sense to rename Godunov to `ScalarGodunovFlux` (https://github.com/mfem/mfem/pull/4513/commits/1e4f2b5d2bc5da72b24241670d613b1f9a63ba46).